Class initialization deadlock causes test timeouts under parallel execution

## Description

During development of PR #514, we observed intermittent test timeouts affecting 18 tests in the core module. These timeouts appear to be environmental rather than code-related, as they:

1. Do not occur when running specific test subsets
2. Affect tests unrelated to the code changes
3. Pass when run individually or in small batches
4. Occur consistently when running full test suite

## Affected Tests

The following tests timeout after 60 seconds:

### Module Loading/Parsing
- `AbstractRecursionPreventingNodeItemVisitorTest.testRecursion`
- `RecursionCollectingNodeItemVisitorTest.testAssemblyRecursion`
- `ExternalConstraintsModulePostProcessorTest.test`
- `ExamplesTest.testExamineAssemblyDefinitionByName`
- `ExamplesTest.testLoadMetaschema`
- `MetaConstraintLoaderTest.test`
- `MetaschemaModuleTest.testFile`
- `ModuleLoaderTest.testConstraints`
- `ModuleLoaderTest.testFile`
- `ModuleLoaderTest.testLoadMetaschemaWithExternalEntity`
- `ModuleLoaderTest.testUrl`

### XML Parsing
- `XmlObjectParserTest.test`
- `XmlbeansMarkupVisitorTest.testComplex`
- `XmlbeansMarkupVisitorTest.testLink`
- `XmlbeansMarkupVisitorTest.testQuote`
- `XmlbeansMarkupVisitorTest.testText`

### Other
- `ModelFactoryTest.letTest`
- `MermaidErDiagramGeneratorTest.testErDiagram`

## Environment

- Platform: Windows (MINGW64_NT-10.0)
- Java: JDK 17 (Temurin)
- Maven: Using toolchains with parallel test execution enabled

## Evidence

### Baseline (develop branch)
Tests pass without timeouts on the develop branch.

### With PR changes
Running numeric-related tests only: **All 159 tests pass**
```
mvn -pl core test -Dtest="**/cst/math/*Test,**/*Numeric*Test,**/*Decimal*Test,**/*Integer*Test"
```

Running full test suite: **18 timeout errors** (all in unrelated tests)

### Dump file
Surefire generates dump files with repeated error: `Index: 2, Size: 1`

## Potential Causes

1. **Parallel test execution conflicts**: Tests may be competing for resources when run in parallel
2. **Windows-specific file locking**: Module loading tests may encounter file locking issues
3. **Resource exhaustion**: Full test suite may exhaust thread pool or memory

## Suggested Investigation

1. Run tests with `-DforkCount=1` to disable parallelism
2. Check for any file handle leaks in module loading
3. Profile test execution to identify resource bottlenecks
4. Compare with Linux/macOS CI behavior

## Investigation Findings

After extensive testing, we identified the root cause of the test timeouts.

### Root Cause: Class Initialization Deadlock

The timeouts are caused by a **class initialization deadlock** in `AbstractArrayItem.java`. When JUnit 5 runs tests in `SEPARATE_THREAD` timeout mode (our default), the separate timeout-monitoring thread can trigger class initialization that deadlocks with the main test thread.

**Stack trace evidence** from `ArraySizeTest.txt`:
```
Caused by: org.junit.jupiter.api.extension.ExecutionTimeoutException: 
Execution timed out in thread junit-timeout-thread-3
  at gov.nist.secauto.metaschema.core.metapath.item.function.AbstractArrayItem.<clinit>(AbstractArrayItem.java:43)
  at gov.nist.secauto.metaschema.core.metapath.item.function.IArrayItem.empty(IArrayItem.java:69)
```

### The Deadlock Pattern

In `AbstractArrayItem.java` (lines 37-43), static field initialization creates a circular dependency chain:

```java
private static final List<IArgument> ARGUMENTS = ObjectUtils.notNull(List.of(
    IArgument.builder().name("position").type(IIntegerItem.type()).one().build()));

private static final IArrayItem<?> EMPTY = new ArrayItemN<>();
```

1. `ARGUMENTS` initialization calls `IIntegerItem.type()`
2. This triggers `IIntegerItem` class initialization
3. Multiple threads waiting for class initialization = deadlock

When JUnit 5's timeout thread checks on a test, it may trigger this initialization chain simultaneously with the test thread, causing both to wait indefinitely for the class lock.

### Reproduction Steps

```bash
mvn test -pl core -DforkCount=4 -DreuseForks=false
```

Consistently reproduces timeouts in:
- `ArraySizeTest`
- `MapContainsTest`  
- `MapSizeTest`

### Relationship to IndexOutOfBoundsException

The `IndexOutOfBoundsException: Index: 1, Size: 1` errors in the tree reporter are a **secondary symptom**. When tests timeout due to the deadlock, the tree reporter encounters inconsistent state when trying to print results, causing the index errors. PR #516 upgraded the tree reporter which may handle this edge case better, but doesn't fix the underlying deadlock.

### Potential Fixes

1. **Lazy initialization**: Use `Lazy<>` pattern for static fields that trigger cross-class initialization
2. **Break circular dependencies**: Restructure initialization order in `AbstractArrayItem`
3. **Change timeout mode**: Use `SAME_THREAD` instead of `SEPARATE_THREAD` (workaround, not a fix)
4. **Eager initialization**: Force class loading order during startup

### Recommended Approach

Convert the problematic static fields to use lazy initialization:

```java
private static final Lazy<List<IArgument>> ARGUMENTS = Lazy.of(() -> 
    ObjectUtils.notNull(List.of(
        IArgument.builder().name("position").type(IIntegerItem.type()).one().build())));

private static final Lazy<IArrayItem<?>> EMPTY = Lazy.of(ArrayItemN::new);
```

This prevents the deadlock by deferring initialization until first use within a single thread context.

--- a/core/src/main/java/gov/nist/secauto/metaschema/core/metapath/item/function/impl/AbstractArrayItem.java
+++ b/core/src/main/java/gov/nist/secauto/metaschema/core/metapath/item/function/impl/AbstractArrayItem.java
@@ -22,6 +22,7 @@ import java.util.Objects;
 import java.util.stream.Collectors;
 
 import edu.umd.cs.findbugs.annotations.NonNull;
+import nl.talsmasoftware.lazy4j.Lazy;
 
 /**
  * The base class for {@link IArrayItem} implementations, that provides an
@@ -35,12 +36,19 @@ public abstract class AbstractArrayItem<ITEM extends ICollectionValue>
     implements IArrayItem<ITEM>, IFeatureCollectionFunctionItem {
   @NonNull
   private static final IEnhancedQName QNAME = IEnhancedQName.of("array");
+  /**
+   * The function arguments, lazily initialized to prevent class initialization
+   * deadlock when multiple threads trigger class loading simultaneously.
+   */
   @NonNull
-  private static final List<IArgument> ARGUMENTS = ObjectUtils.notNull(List.of(
-      IArgument.builder().name("position").type(IIntegerItem.type()).one().build()));
-
+  private static final Lazy<List<IArgument>> ARGUMENTS = ObjectUtils.notNull(Lazy.of(() -> ObjectUtils.notNull(List.of(
+      IArgument.builder().name("position").type(IIntegerItem.type()).one().build()))));
+  /**
+   * An empty array item singleton, lazily initialized to prevent class
+   * initialization deadlock.
+   */
   @NonNull
-  private static final IArrayItem<?> EMPTY = new ArrayItemN<>();
+  private static final Lazy<IArrayItem<?>> EMPTY = ObjectUtils.notNull(Lazy.of(ArrayItemN::new));
 
   /**
    * Get an immutable array item that is empty.
@@ -52,7 +60,7 @@ public abstract class AbstractArrayItem<ITEM extends ICollectionValue>
   @SuppressWarnings("unchecked")
   @NonNull
   public static <T extends ICollectionValue> IArrayItem<T> empty() {
-    return (IArrayItem<T>) EMPTY;
+    return (IArrayItem<T>) EMPTY.get();
   }
 
   @Override
@@ -62,7 +70,7 @@ public abstract class AbstractArrayItem<ITEM extends ICollectionValue>
 
   @Override
   public List<IArgument> getArguments() {
-    return ARGUMENTS;
+    return ARGUMENTS.get();
   }
 
   @Override

--- a/core/src/main/java/gov/nist/secauto/metaschema/core/metapath/item/function/impl/AbstractMapItem.java
+++ b/core/src/main/java/gov/nist/secauto/metaschema/core/metapath/item/function/impl/AbstractMapItem.java
@@ -26,6 +26,7 @@ import java.util.Objects;
 import java.util.stream.Collectors;
 
 import edu.umd.cs.findbugs.annotations.NonNull;
+import nl.talsmasoftware.lazy4j.Lazy;
 
 /**
  * The base class for {@link IMapItem} implementations, that provide an
@@ -43,13 +44,18 @@ public abstract class AbstractMapItem<VALUE extends ICollectionValue>
   @NonNull
   private static final IEnhancedQName QNAME = IEnhancedQName.of("map");
   /**
-   * The function arguments.
+   * The function arguments, lazily initialized to prevent class initialization
+   * deadlock when multiple threads trigger class loading simultaneously.
    */
   @NonNull
-  private static final List<IArgument> ARGUMENTS = ObjectUtils.notNull(List.of(
-      IArgument.builder().name("key").type(IAnyAtomicItem.type()).one().build()));
+  private static final Lazy<List<IArgument>> ARGUMENTS = ObjectUtils.notNull(Lazy.of(() -> ObjectUtils.notNull(List.of(
+      IArgument.builder().name("key").type(IAnyAtomicItem.type()).one().build()))));
+  /**
+   * An empty map item singleton, lazily initialized to prevent class
+   * initialization deadlock.
+   */
   @NonNull
-  private static final IMapItem<?> EMPTY = new MapItemN<>();
+  private static final Lazy<IMapItem<?>> EMPTY = ObjectUtils.notNull(Lazy.of(MapItemN::new));
 
   /**
    * Get an immutable map item that is empty.
@@ -62,7 +68,7 @@ public abstract class AbstractMapItem<VALUE extends ICollectionValue>
   @SuppressWarnings("unchecked")
   @NonNull
   public static <V extends ICollectionValue> IMapItem<V> empty() {
-    return (IMapItem<V>) EMPTY;
+    return (IMapItem<V>) EMPTY.get();
   }
 
   @Override
@@ -72,7 +78,7 @@ public abstract class AbstractMapItem<VALUE extends ICollectionValue>
 
   @Override
   public List<IArgument> getArguments() {
-    return ARGUMENTS;
+    return ARGUMENTS.get();
   }
 
   @Override

